### ISSUE TITLE
Are you planning to implement Apollo Federation?

### ISSUE THREAD

**Issue by voslartomas** [19majkel94/type-graphql#351]
**Is your feature request related to a problem? Please describe.**
I guess it would be about to add some decorators for @key, @extend etc, to be able to declare federation schema services.

**Describe the solution you'd like**
https://blog.apollographql.com/apollo-federation-f260cf525d21

**Comment by 19majkel94** [comment]
> I guess it would be about to add some decorators for @key, @extend etc

So it's blocked by #77 😞 
But when #77 is ready, you would be able to use Apollo Federation with TypeGraphQL without any problem, so no further integration needed 🔒

**Comment by voslartomas** [comment]
@19majkel94 I see, I've read the issue and saw that you agreed (february) with guys from graphql-js, but not sure what does that mean for current situation? Is it going to be blocked, or something is happening?

**Comment by 19majkel94** [comment]
I am waiting for changes in `graphql-js` and interchangeability between `extensions` and `directives`, so both approaches (SDL and code-first) could be used and handled by universal directives and other tools like prisma or apollo federation.

And @Hossein-s is working on a workaround to modify `astNodes` that are currently used by all JS tools to consume directives metadata from SDL.

**Comment by 19majkel94** [comment]
> I guess it would be about to add some decorators for @key, @extend etc, 

Looks like it's not only about directives but also the `extend` keyword, which is handled in #228 🔑

**Comment by j** [comment]
@19majkel94 is it the extend keyword in #228 though?  Doesn't #228 do the whole, `type User extends BaseType`.. The `extend` keyword in federation goes before the type, it's a way they can make the root graphql service have types that each field calls different services.

**Comment by 19majkel94** [comment]
@j I know, in #228 I was thinking about both merging typedefs from different file and `extend type Foo` syntax if it's possible with `graphql-js`.

**Comment by j** [comment]
@19majkel94 bummer, yeah looks like most "schema-last" node.js implementations are going to have a rough time from what I was reading.  I saw a post of someone requesting a different way of making schemas to make it easier, but I'm not sure what it'd take.  Apollo Federation is the next step for our company's infrastructure and will have to move away from type-graphql if support doesn't come to, and that'll be a sad day.  :(

**Comment by j** [comment]
https://github.com/prisma/nexus/issues/148 for reference

**Comment by 19majkel94** [comment]
Maybe I could just drop building schema using `graphql-js` and create the SDL string on my own (like in the reflection plugin PoC), and then just plug in the resolvers using `graphql-tools` to create an executable schema if it's needed (or not in the case of the type-graphql core.

**Comment by j** [comment]
@19majkel94 that sounds pretty flexible?

**Comment by j** [comment]
Another issue to help solve code first libraries: https://github.com/apollographql/apollo-server/issues/2769

**Comment by j** [comment]
Another note I saw regarding "extends".  https://www.apollographql.com/docs/apollo-server/federation/federation-spec/

The spec actually supports this:

```graphql
type User @key(fields: "id") @extends {
  id: ID! @external
  reviews: [Review]
}
```

But still looks like a lot of work, :(

**Comment by MichalLytek** [comment]
I've added an example of apollo-federation (a modified federation demo from @j PR):
https://github.com/MichalLytek/type-graphql/tree/20c81f4f7ee82779595002d25784fce3a8ff8b9b/examples/apollo-federation

I will leave this PR open to collect some feedback about the integration.
Also in the future I plan to make `@typegraphql/federation` package with `@FederationKey` decorator and others 😉

**Comment by lgabeskiria** [comment]
Hi guys! Thanks for this amazing library! 
Do you have any ETA when can we expect this feature in production?

**Comment by MichalLytek** [comment]
@lgabeskiria Have you tried the latest beta release? Without testing and confirmation I won't release "stable" package.

**Comment by lgabeskiria** [comment]
@MichalLytek will try to provide a feedback this week, thanks!

**Comment by Sector95** [comment]
Got it working in our (currently) very simple project today, so far so good...  Will continue working with it over the next couple of days and will report back as well.

**Comment by lgabeskiria** [comment]
@MichalLytek tested the feature, looks good to me!

**Comment by lgabeskiria** [comment]
Hi @MichalLytek is it possible to release this feature as beta release?

**Comment by MichalLytek** [comment]
@lgabeskiria 
Have you checked the npm site and typed `npm i type-graphql@beta`?

**Comment by tuxmachine** [comment]
Tested the beta release with the GraphQL module in nestjs and works like a charm.

**Comment by adamovittorio** [comment]
Hi All, I've installed the `beta`, used the same approach described in the example with [buildFederatedSchema](https://github.com/MichalLytek/type-graphql/blob/20c81f4f7ee82779595002d25784fce3a8ff8b9b/examples/apollo-federation/helpers/buildFederatedSchema.ts) and copying in my source code [createResolversMap](https://github.com/MichalLytek/type-graphql/blob/20c81f4f7ee82779595002d25784fce3a8ff8b9b/src/utils/createResolversMap.ts)
and his dependencies because they are not exported from `type-graphql`.

It seems to work to me but the schema generated doesn't have the `@key` annotation.

**player.ts**
```
import { ObjectType, Field, ID, Directive } from "type-graphql";

@Directive(`@key(fields: "id")`)
@ObjectType()
export default class Player {
  @Field(_type => ID)
  id: string;

  @Field()
  username: string;
}

```

**resolver.ts**

```
import { Query, Resolver } from "type-graphql";
import Player from "./player";

@Resolver(_of => Player)
export default class PlayerResolver {
  player = {
    id: "123",
    username: "victorious",
  };

  @Query(_returns => Player)
  async me(): Promise<Player> {
    return this.player;
  }
}
```
**schema.graphql**

```
# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!) on INTERFACE | OBJECT

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

type Player {
  id: ID!
  username: String!
}

type Query {
  me: Player!
}

```

Am I doing something wrong?

**Comment by MichalLytek** [comment]
@adamovittorio
> It seems to work to me but the schema generated doesn't have the @key annotation.

![image](https://user-images.githubusercontent.com/10618781/70065832-02b36d80-15ec-11ea-8b1a-b4143dc828bf.png)

https://typegraphql.ml/docs/next/directives.html

**Comment by adamovittorio** [comment]
@MichalLytek - Thank you very much, I miss that part 😅.

Regarding `buildFederatedSchema`, will it be exposed from `type-graphql`/`@typegraphql/federation`?

**Comment by MichalLytek** [comment]
In the future yes, in `@typegraphql/federation` I think 😉

**Comment by adamovittorio** [comment]
Hi @MichalLytek, do you have any update regarding `@typegraphql/federation`, can I help in some way?

**Comment by Sector95** [comment]
Is #521 expected to change how this is implemented?

ie. Will we be using the @Extensions annotation in some way instead?

**Comment by MichalLytek** [comment]
@Sector95 In the future yes 😉 Apollo team has to expose a way to build federated schema from existing `GraphQLSchema` object and use `extensions` instead of AST directives:
https://github.com/apollographql/apollo-server/pull/3013

**Comment by SamXDesc** [comment]
> Tested the beta release with the GraphQL module in nestjs and works like a charm.

Can you provide a demo repository for us?

**Comment by tuxmachine** [comment]
> Can you provide a demo repository for us?

There's a small demo in my fork of the NestJS GraphQL package (PR still open)

https://github.com/tuxmachine/graphql/tree/feat/enable-code-first/tests/type-graphql-federation

**Comment by graysonhicks** [comment]
Is there an open PR for this feature or has it already been merged somewhere?

**Comment by tuxmachine** [comment]
> Is there an open PR for this feature or has it already been merged somewhere?

 The PR currently supports federation using regular schemas, type-graphql support is included but disabled. It can be enabled once the federation support in this repository is no longer beta.

https://github.com/nestjs/graphql/pull/455

**Comment by david-golightly-leapyear** [comment]
@tuxmachine It appears that Nest.js has merged and published the linked PR (which you're the author of!). Is there an update on enabling type-graphql support?

**Comment by gperdomor** [comment]
Hi... Same question here... Any update about enabling type-graphql support?

**Comment by tuxmachine** [comment]
No updates. It can be enabled once the federation support in this repository is no longer beta.

Currently it would require `type-graphql` as a beta dependency. Tell me if I'm missing the point, but it seems strange for a _stable_ release of `@nestjs/graphql` to have a _beta_ dependency. If you can make a good argument for it, the fix is as simple as removing [this line](https://github.com/nestjs/graphql/blob/master/lib/graphql-federation.factory.ts#L34)

**Comment by gperdomor** [comment]
@tuxmachine well, if we consider beta anything with version minor to 1.0.0, then TypeORM is also a beta... and as we know, is one of the best ORM for typescript 🤔

**Comment by tuxmachine** [comment]
@gperdomor No, I consider it beta because it's listed under "unreleased" in the [changelogs of `type-graphql`](https://github.com/MichalLytek/type-graphql/blob/master/CHANGELOG.md#features), and the code enabling Directives is packaged inside `type-graphql@0.18.0-beta.11`

**Comment by gperdomor** [comment]
ohh I see... In that case I think we should wait...

**Comment by henry-young** [comment]
I've been playing around with the [typegrahpql federation example](https://github.com/MichalLytek/type-graphql/tree/master/examples/apollo-federation) and I don't know how to get the auth checker to validate fields decorated with`@Authorized()` when using `__resolveReference`. Is there a way to trigger the auth checks manually from the `resolve{Type}Reference()` function?

**Comment by bsparks** [comment]
@MichalLytek  I have been looking into using the federation example now with rc1, I think the createResolversMap util method needs to be exposed externally in order to use it outside this repo? https://github.com/MichalLytek/type-graphql/blob/master/examples/apollo-federation/helpers/buildFederatedSchema.ts#L10

**Comment by MichalLytek** [comment]
@bsparks done in `7ddface` 💪

**Comment by andrei-bitca-dc** [comment]
Hello,

  I ran into a problem when trying to build the federation schema using `buildFederatedSchema` example - It seems that when defining default values for an input enum type property and the enum resolver has different values than keys, it doesn't recognize the default enum value:

**Example:**

Schema:

```
enum TestEnum {
   A
}

input TestInput {
  prop: TestEnum = A
}
```

TestEnum resolver:

```
enum TestEnum {
  A = "some_different_value_than_A",
}

registerEnumType(TestEnum, {
  name: "TestEnum"
});
```

**Error on example:**

GraphQLError: Enum "TestEnum" cannot represent value: "A"

**Possible solution:**

Changing https://github.com/MichalLytek/type-graphql/blob/e32a34fe065745bd209eadc5c36ba8bef6d893f4/src/utils/createResolversMap.ts#L41 to:

```
enumMap[name] = name;
```

fixed it.

**Comment by MichalLytek** [comment]
@andrei-bitca-dc 
It cannot be changed to `enumMap[name] = name` because enum values are runtime value and enum keys are enum names in schema.

This is the behavior of the official GraphQL implementation, so you should use internal values as default values:
https://github.com/graphql/graphql-js/issues/1604

**Comment by andrei-bitca-dc** [comment]
The problem is that it throws me this error: `GraphQLError: Enum "TestEnum" cannot represent value: "A"` with the following InputType definition:

```
@InputType()
export class TestInput {
  @Field(_type => EnumType, { defaultValue: EnumType.A })
  public prop: EnumType = EnumType.A;
}
```

**Comment by MichalLytek** [comment]
@andrei-bitca-dc Can you reproduce that with the normal `buildSchema`? Maybe apollo federation is doing something wrong or they have tried to "fix" that issue

**Comment by andrei-bitca-dc** [comment]
@MichalLytek `buildSchemaSync` works.

**Comment by MichalLytek** [comment]
So please try to reproduce the case with raw `graphql-js` and raise the issue on theirs side 😉

**Comment by andrei-bitca-dc** [comment]
@MichalLytek FYI: https://github.com/apollographql/apollo-server/issues/4143.

**Comment by yaacovCR** [comment]
GraphQL-Tools's schema stitching now supports type merging similar to Federation, as well as specification  of merge instructions for the gateway via directives (see our docs and https://github.com/gmac/schema-stitching-handbook for examples).

We want to support specification of merge instructions via extensions for code first schemas such as TypeGraphQL Just checking first if TypeGraphQL lets you specify extensions for types and fields, and want to inquire as to whether there is anything I should know about how you handle extensions within the framework. I imagine that we will follow supposed best practices and nest our type field extensions under a GRAPHQL_TOOLS_SCHEMA_STITCHING key... We will probably have options on the gateway to modify the location of the extensions as desired. Just inquiring whether there is any TypeGraphQL specific info about extensions that I should be aware of...

**Comment by MichalLytek** [comment]
@yaacovCR https://typegraphql.com/docs/extensions.html

**Comment by yaacovCR** [comment]
Looks pretty straightforward, thanks!

**Comment by yaacovCR** [comment]
Decided to go a different way and follow Gatsby/graphql-compose convention of reading directives from `extensions.directives` -- see https://github.com/graphql/graphql-js/issues/1343#issuecomment-479871020 -- but I also allowed this to be customized as desired.

This is now released as canary, see https://github.com/ardatan/graphql-tools/pull/2391

To get this working:
1. Add the prebuilt directives to the schema from `@graphql-tools/stitching-directives`
2. Use the directives within the schema via `extensions.directives`
3. Expose the SDL to the gateway via some resolver (_root, _args, _context, info) => printSchemaWithDirectives(info.schema) where printSchemaWithDirectives is an improved `@graphql-tools/utils` function that follows same convention as above.

For a peak at fully worked out examples of stitching with SDL-first approaches, take a look at https://github.com/gmac/schema-stitching-handbook, where @gmac takes you through step-by-step from the basics all the way to hot reloading/versioning releases. 

Hopefully, soon an example will appear with code-first approach.

Thanks!

**Comment by yaacovCR** [comment]
Nexus support is live with stitching directives (in extensions):

https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/products/schema.js

note that although TypeGraphQL supports tagging schema entities (types, fields) with directive nodes (directive use on schema entities) either directly or via extensions, it does not seem to support adding actual directive entities (the actual definitions of custom directives).

I think this is incorrect behavior, a server might want to implement a custom directive on a query.

You can do this quite easily in `graphql-js`
```
new GraphQLSchema({
  ...
  directives: [...specifiedDirectives, ...your_array_of_custom_directives],
})
```

I worked around this by using `extendSchema` to add in more typeDefs, but see the `graphql-js` example where I just loaded in the actual directives. (https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/accounts/schema.js)

**Comment by yaacovCR** [comment]
TypeGraphQL support is live with stitching directives:

https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/inventory/schema.ts

note that although TypeGraphQL supports tagging schema entities (types, fields) with directive nodes (directive use on schema entities) via extensions, it does not seem to support adding actual directive entities (the actual definitions of custom directives).

I think this is incorrect behavior, a server might want to implement a custom directive on a query.

You can do this quite easily in `graphql-js`
```
new GraphQLSchema({
  ...
  directives: [...specifiedDirectives, ...your_array_of_custom_directives],
})
```

I worked around this by using `extendSchema` to add in more typeDefs, but see the `graphql-js` example where I just loaded in the actual directives. (https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/accounts/schema.js)

**Comment by yaacovCR** [comment]
TypeGraphQL support is live with stitching directives:

https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/products/schema.ts

note that although TypeGraphQL supports tagging schema entities (types, fields) with directive nodes (directive use on schema entities) via extensions, it does not seem to support adding actual directive entities (the actual definitions of custom directives).

I think this is incorrect behavior, a server might want to implement a custom directive on a query.

You can do this quite easily in `graphql-js`
```
new GraphQLSchema({
  ...
  directives: [...specifiedDirectives, ...your_array_of_custom_directives],
})
```

I worked around this by using `extendSchema` to add in more typeDefs, but see the `graphql-js` example where I just loaded in the actual directives. (https://github.com/yaacovCR/schema-stitching-demos/blob/code-first/code-first-schemas/services/accounts/schema.js)

**Comment by MichalLytek** [comment]
> it does not seem to support adding actual directive entities (the actual definitions of custom directives).

I'm not sure if I understand that correctly but I'm sure it does:

![image](https://user-images.githubusercontent.com/10618781/103135573-85a4db80-46b9-11eb-82f8-f2ddabb46d17.png)

**Comment by yaacovCR** [comment]
Woops, fantastic!

**Comment by yaacovCR** [comment]
Example updated. We don't need skipCheck true even!

**Comment by vongohren** [comment]
So one cannnot use type graphql with apollo federation? Is it impossible or just tricky?

**Comment by MichalLytek** [comment]
@vongohren https://github.com/MichalLytek/type-graphql/issues/351#issuecomment-549156066

**Comment by tim-trisnadhama** [comment]
> I've added an example of apollo-federation (a modified federation demo from @j PR):
> https://github.com/MichalLytek/type-graphql/tree/20c81f4f7ee82779595002d25784fce3a8ff8b9b/examples/apollo-federation
> 
> I will leave this PR open to collect some feedback about the integration.
> Also in the future I plan to make `@typegraphql/federation` package with `@FederationKey` decorator and others like defining `resolverReference`, etc. 😉

Hi, is there any update on the `@FederationKey`? I m having trouble getting the context from the resolveReference since we inject the context when we initialize the server

**Comment by tim-trisnadhama** [comment]
I got it to work by declaring it as a `const`

```
export const UserReferenceResolver = async (ref: Pick<User, 'Id'>, ctx: any, info: any): Promise<User> => {
  const fields = info.fieldNodes[0].selectionSet.selections[0].selectionSet.selections.map(
    (field: { name: { value: any } }) => field.name.value,
  );
  const option = { filters: { Id: ref.Id } };
  const user: User[] = await userRepo.getMany(ctx, fields, option as any);
  return user[0];
};
```

Then when building the federated schema, i just add this to the `schema`:
```
const schema = await buildFederatedSchema(
    {
      emitSchemaFile: true,
      nullableByDefault: true,
      resolvers,
      validate: false,
    },
    {
      User: { __resolveReference: UserReferenceResolver },
      Account: { __resolveReference: AccountReferenceResolver },
    },
  );
```

I guess this is the only workaround until `@FederationKey` is implemented?

**Comment by joemckie** [comment]
I found a much nicer way to do this using `buildSubgraphSchema` from `@apollo/federation`:

```ts
const { typeDefs, resolvers } = await buildTypeDefsAndResolvers({
  resolvers: [...],
  orphanedTypes: [...],
});

const schema = buildSubgraphSchema({
  typeDefs: gql(typeDefs),
  resolvers: resolvers as GraphQLResolverMap
});
```

**Comment by biancazzurri** [comment]
> I got it to work by declaring it as a `const`
> 
> ```
> export const UserReferenceResolver = async (ref: Pick<User, 'Id'>, ctx: any, info: any): Promise<User> => {
>   const fields = info.fieldNodes[0].selectionSet.selections[0].selectionSet.selections.map(
>     (field: { name: { value: any } }) => field.name.value,
>   );
>   const option = { filters: { Id: ref.Id } };
>   const user: User[] = await userRepo.getMany(ctx, fields, option as any);
>   return user[0];
> };
> ```
> 
> Then when building the federated schema, i just add this to the `schema`:
> 
> ```
> const schema = await buildFederatedSchema(
>     {
>       emitSchemaFile: true,
>       nullableByDefault: true,
>       resolvers,
>       validate: false,
>     },
>     {
>       User: { __resolveReference: UserReferenceResolver },
>       Account: { __resolveReference: AccountReferenceResolver },
>     },
>   );
> ```
> 
> I guess this is the only workaround until `@FederationKey` is implemented?

this works for me too! thanks

How hard is it to create syntactic sugar for this via @ObjectType directive, and make it receive __resolveReference function?
